sklearn/datasets/_california_housing.py

```diff
@@ -155,10 +171,23 @@ def fetch_california_housing(data_home=None, download_if_missing=True,
     with open(join(module_path, 'descr', 'california_housing.rst')) as dfile:
         descr = dfile.read()
 
+    X = data
+    y = target
     if return_X_y:
-        return data, target
-
-    return Bunch(data=data,
-                 target=target,
+        return X, y
+
+    frame = None
+    target_names = ["MedHouseVal", ]
+    if as_frame:
+        columns = feature_names + target_names
+        adjusted_data = np.hstack((data, target[:,np.newaxis]))
```

Comment:
This is not a good idea in case X and y have different data types. numpy doesn't allow different types, so you're making ``y`` be float here (if it was int) or ``X`` be object (if y is object).

And I think the approach of wrapping this in a reusable function is good.

Reply:
The current implementation with numpy worked for this specific dataset, but I agree that it wouldn't generalize well. 
 I'll try making this more generalizable -- maybe by constructing separate dataframes (one for `X` and one for `y`), and using pandas to concatenate them together instead of numpy.  🤔

Reply:
yes something like that would probably work well. Have you tried working with  @wconnell as well?

Reply:
Cool!  I just made that proposed change (in the latest commit below) and tested it locally.  The `_convert_data_dataframe` method in this PR hopefully will be able to generalize to some other cases.

I haven't tried working with @wconnell but am open to it!  Will try to find them on gitter.